test/e2e/contractOrdersWithGnosisSafe.test.ts

```diff
@@ -0,0 +1,286 @@
+  let deployer: Wallet;
+  let solver: Wallet;
+  let trader: Wallet;
+  let safeOwners: Wallet[];
+
+  let settlement: Contract;
+  let allowanceManager: Contract;
+  let safe: Contract;
+  let domainSeparator: TypedDataDomain;
+  let gnosisSafeManager: GnosisSafeManager;
+  let GnosisSafeEIP1271Fallback: ContractFactory;
+
+  beforeEach(async () => {
+    const deployment = await deployTestContracts();
+
+    ({
+      deployer,
+      settlement,
+      allowanceManager,
+      wallets: [solver, trader, ...safeOwners],
+    } = deployment);
+
+    const { authenticator, owner } = deployment;
+    await authenticator.connect(owner).addSolver(solver.address);
+
+    const { chainId } = await ethers.provider.getNetwork();
+    domainSeparator = domain(chainId, settlement.address);
+
+    GnosisSafeEIP1271Fallback = await ethers.getContractFactory(
+      "GnosisSafeEIP1271Fallback",
+    );
+    const fallback = await GnosisSafeEIP1271Fallback.deploy();
+
+    gnosisSafeManager = await GnosisSafeManager.init(deployer);
+    safe = await gnosisSafeManager.newSafe(
+      safeOwners.map((wallet) => wallet.address),
+      2,
+      fallback.address,
+    );
+  });
+
+  it("should settle matching orders", async () => {
+    // EOA trader: sell 1 WETH for  900 DAI
+    // Safe:       buy  1 WETH for 1100 DAI
+    // Settlement price at 1000 DAI for 1 WETH.
+
+    const erc20 = (symbol: string) =>
+      waffle.deployContract(deployer, ERC20, [symbol, 18]);
+
+    const dai = await erc20("DAI");
+    const weth = await erc20("WETH");
+
+    const UNLIMITED_VALID_TO = 0xffffffff;
+    const encoder = new SettlementEncoder(domainSeparator);
+
+    const TRADER_FEE = ethers.utils.parseEther("0.001");
+    const TRADER_SOLD_AMOUNT = ethers.utils.parseEther("1.0");
+    const TRADER_BOUGHT_AMOUNT = ethers.utils.parseEther("900.0");
+
+    await weth.mint(trader.address, TRADER_SOLD_AMOUNT.add(TRADER_FEE));
+    await weth
+      .connect(trader)
+      .approve(allowanceManager.address, ethers.constants.MaxUint256);
+
+    await encoder.signEncodeTrade(
+      {
+        kind: OrderKind.SELL,
+        partiallyFillable: false,
+        sellToken: weth.address,
+        buyToken: dai.address,
+        sellAmount: TRADER_SOLD_AMOUNT,
+        buyAmount: TRADER_BOUGHT_AMOUNT,
+        appData: 0,
+        validTo: UNLIMITED_VALID_TO,
+        feeAmount: TRADER_FEE,
+      },
+      trader,
+      SigningScheme.ETHSIGN,
+    );
+
+    const SAFE_FEE = ethers.utils.parseEther("10.0");
+    const SAFE_SOLD_AMOUNT = ethers.utils.parseEther("1100.0");
+    const SAFE_BOUGHT_AMOUNT = ethers.utils.parseEther("1.0");
+
+    await dai.mint(safe.address, SAFE_SOLD_AMOUNT.add(SAFE_FEE));
+    const approveTransaction = {
+      to: dai.address,
+      data: dai.interface.encodeFunctionData("approve", [
+        allowanceManager.address,
+        ethers.constants.MaxUint256,
+      ]),
+    };
+    await execSafeTransaction(safe, approveTransaction, safeOwners);
```

Comment:
Would it be possible to encode the approve transaction as a pre-interaction to make the trade entirely gasless for the safe owners?

Reply:
Yep, this should be possible. For the test I prefer to leave it explicit just as a small extra test that signing works.